docs/proposals/gpu-support.md

```diff
@@ -0,0 +1,339 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN MUNGE: GENERATED_TOC -->
+
+- [GPU support](#gpu-support)
+  - [Objective](#objective)
+  - [Background](#background)
+  - [Detailed discussion](#detailed-discussion)
+    - [Inventory](#inventory)
+    - [Scheduling](#scheduling)
+    - [The runtime](#the-runtime)
+      - [NVIDIA support](#nvidia-support)
+    - [Event flow](#event-flow)
+  - [Implementation plan](#implementation-plan)
+    - [V0](#v0)
+      - [Scheduling](#scheduling)
+      - [Runtime](#runtime)
+      - [Other](#other)
+  - [Future work](#future-work)
+    - [V1](#v1)
+    - [V2](#v2)
+    - [V3](#v3)
+    - [Undetermined](#undetermined)
+  - [Security considerations](#security-considerations)
+
+<!-- END MUNGE: GENERATED_TOC -->
+
+# GPU support
+
+Author: @therc
+
+Date: Apr 2016
+
+Status: Design in progress, early implementation of requirements
+
+## Objective
+
+Users should be able to request GPU resources for their workloads, as easily as
+for CPU or memory. Kubernetes should keep an inventory of machines with GPU
+hardware, schedule containers on appropriate nodes and set up the container
+environment with all that's necessary to access the GPU. All of this should
+eventually be supported for clusters on either bare metal or cloud providers.
+
+## Background
+
+An increasing number of workloads, such as machine learning and seismic survey
+processing, benefits from offloading computations to graphic hardware. While not
+as tuned as traditional, dedicated high performance computing systems such as
+MPI, a Kubernetes cluster can still be a great environment for organizations
+that need a variety of additional, "classic" workloads, such as database, web
+serving, etc.
+
+GPU support is hard to provide extensively and will thus take time to tame
+completely, because
+
+- different vendors expose the hardware to users in different ways
+- some vendors require fairly tight coupling between the kernel driver
+controlling the GPU and the libraries/applications that access the hardware
+- it adds more resource types (whole GPUs, GPU cores, GPU memory)
```

Comment:
It doesn't qualify as a cluster manager, but our legacy environment packs together multiple processes on the same card through static allocation. At some point I'd like to do something similar with Kubernetes. We can't keep using the legacy pipeline because we're fed up with VMs. :-)  Even assuming exclusive use of the card, some might want to declare "I want to run this on a card with at least x GB". We have some workloads that fit the beefier hardware in our colocated machines, but not on AWS setups. They really only care about the memory.